### PR TITLE
Make sdl2-compat's cmake script even more compatible

### PR DESCRIPTION
**Pull request by madebr**
- Add `-Wl,--no-undefined` if supported by toolchain to catch surprising undefined references
- Build `SDL2.dll` instead of `libSDL2.dll`
- Create `libSDL2_test.a` instead of `libSDL2test.a`
- Create `SDL2::SDL2main` target
- Avoid needing `-lmingw32` by adding `-Wl,-undefined,WinMain`
- Fix buidling testcontrollermap.c + testgamecontroller.c by addint test/testutils.c
- Only build OpenGLES tests when headers are available.
- testnative uses platform dependent sources

Prefix of SDL2 on MinGW looks as:
```
prefix
├── bin
│   ├── sdl2-config
│   └── SDL2.dll
├── include
│   └── SDL2
│       ├── begin_code.h
│       ├── close_code.h
│       ├── SDL_assert.h
│       ├── SDL_atomic.h
│       ├── SDL_audio.h
│       ├── SDL_bits.h
│       ├── SDL_blendmode.h
│       ├── SDL_clipboard.h
│       ├── SDL_config_android.h
│       ├── SDL_config_emscripten.h
│       ├── SDL_config.h
│       ├── SDL_config_iphoneos.h
│       ├── SDL_config_macosx.h
│       ├── SDL_config_minimal.h
│       ├── SDL_config_ngage.h
│       ├── SDL_config_os2.h
│       ├── SDL_config_pandora.h
│       ├── SDL_config_windows.h
│       ├── SDL_config_wingdk.h
│       ├── SDL_config_winrt.h
│       ├── SDL_config_xbox.h
│       ├── SDL_copying.h
│       ├── SDL_cpuinfo.h
│       ├── SDL_egl.h
│       ├── SDL_endian.h
│       ├── SDL_error.h
│       ├── SDL_events.h
│       ├── SDL_filesystem.h
│       ├── SDL_gamecontroller.h
│       ├── SDL_gesture.h
│       ├── SDL_guid.h
│       ├── SDL.h
│       ├── SDL_haptic.h
│       ├── SDL_hidapi.h
│       ├── SDL_hints.h
│       ├── SDL_joystick.h
│       ├── SDL_keyboard.h
│       ├── SDL_keycode.h
│       ├── SDL_loadso.h
│       ├── SDL_locale.h
│       ├── SDL_log.h
│       ├── SDL_main.h
│       ├── SDL_messagebox.h
│       ├── SDL_metal.h
│       ├── SDL_misc.h
│       ├── SDL_mouse.h
│       ├── SDL_mutex.h
│       ├── SDL_name.h
│       ├── SDL_opengles2_gl2ext.h
│       ├── SDL_opengles2_gl2.h
│       ├── SDL_opengles2_gl2platform.h
│       ├── SDL_opengles2.h
│       ├── SDL_opengles2_khrplatform.h
│       ├── SDL_opengles.h
│       ├── SDL_opengl_glext.h
│       ├── SDL_opengl.h
│       ├── SDL_pixels.h
│       ├── SDL_platform.h
│       ├── SDL_power.h
│       ├── SDL_quit.h
│       ├── SDL_rect.h
│       ├── SDL_render.h
│       ├── SDL_revision.h
│       ├── SDL_rwops.h
│       ├── SDL_scancode.h
│       ├── SDL_sensor.h
│       ├── SDL_shape.h
│       ├── SDL_stdinc.h
│       ├── SDL_surface.h
│       ├── SDL_system.h
│       ├── SDL_syswm.h
│       ├── SDL_test_assert.h
│       ├── SDL_test_common.h
│       ├── SDL_test_compare.h
│       ├── SDL_test_crc32.h
│       ├── SDL_test_font.h
│       ├── SDL_test_fuzzer.h
│       ├── SDL_test.h
│       ├── SDL_test_harness.h
│       ├── SDL_test_images.h
│       ├── SDL_test_log.h
│       ├── SDL_test_md5.h
│       ├── SDL_test_memory.h
│       ├── SDL_test_random.h
│       ├── SDL_thread.h
│       ├── SDL_timer.h
│       ├── SDL_touch.h
│       ├── SDL_types.h
│       ├── SDL_version.h
│       ├── SDL_video.h
│       └── SDL_vulkan.h
├── lib
│   ├── cmake
│   │   └── SDL2
│   │       ├── SDL2Config.cmake
│   │       ├── SDL2ConfigVersion.cmake
│   │       ├── SDL2mainTargets.cmake
│   │       ├── SDL2mainTargets-debug.cmake
│   │       ├── SDL2Targets.cmake
│   │       ├── SDL2Targets-debug.cmake
│   │       ├── SDL2_testTargets.cmake
│   │       └── SDL2_testTargets-debug.cmake
│   ├── libSDL2.dll.a
│   ├── libSDL2main.a
│   ├── libSDL2_test.a
│   └── pkgconfig
│       └── sdl2_compat.pc
└── share
    └── aclocal
        └── sdl2.m4
```

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -209,11 +222,19 @@ endif()
 #else()
     add_library(SDL2main STATIC src/SDLmain/dummy/SDL_dummy_main.c)
 #endif()
-#add_library(SDL2::SDL2ain ALIAS SDL2main)
+add_library(SDL2::SDL2main ALIAS SDL2main)
 target_include_directories(SDL2main PRIVATE "include/SDL2")
+if(MINGW OR CYGWIN)
+  cmake_minimum_required(VERSION 3.13)
```
Comment (sezero) [inline CMakeLists.txt:228]: Do we have a better (another) way for this than requiring cmake >= 3.13?

File: CMakeLists.txt [hunk 2]
```diff
@@ -22,7 +22,20 @@ include(CheckCSourceCompiles)
...
 include(CheckCCompilerFlag)
 include(CMakePackageConfigHelpers)
+include(CMakePushCheckState)
 include(GNUInstallDirs)
+include("${CMAKE_CURRENT_LIST_DIR}/cmake/sdlchecks.cmake")
+
+CheckOpenGLES()
+
+cmake_push_check_state()
+set(CMAKE_REQUIRED_FLAGS "-Wl,--no-undefined")
```
Comment (sezero) [inline CMakeLists.txt:32]: Note: SDL autofoo and cmake'ry blacklists OpenBSD for this.

File: CMakeLists.txt (hunk 1 above)
Comment (madebr) [reply CMakeLists.txt:228]: Oh, this can be removed. It was `target_link_options` before, as done in SDL itself.

Comment (sezero) [reply CMakeLists.txt]: SDL(2) itself has the same thing, sadly.

File: CMakeLists.txt (hunk 2 above)
Comment (madebr) [reply CMakeLists.txt:32]: I found this: https://github.com/libsdl-org/SDL/issues/3720
So let's disable it there again.

File: CMakeLists.txt (hunk 1 above)
Comment (madebr) [reply CMakeLists.txt]: I think we can get away with `target_link_libraries`.
On the other hand, nobody complained that it didn't work for their older CMake.
So everybody was able to upgrade.

Comment (sezero) [reply CMakeLists.txt]: > On the other hand, nobody complained that it didn't work for their older CMake.

I did :))) I'm still with 3.9.6 on my everyday machine, can't build a newer one.

Comment (madebr) [reply CMakeLists.txt]: How do you build SDL_image with CMake then?

Comment (sezero) [reply CMakeLists.txt]: > How do you build SDL_image with CMake then?

I don't, I just use autotools.

File: CMakeLists.txt [hunk 3]
```diff
@@ -347,8 +371,10 @@ if(SDL2COMPAT_TESTS)
   test_program(torturethread "test/torturethread.c")
   test_program(testgl2 "test/testgl2.c")
   test_program(testgles "test/testgles.c")
-  test_program(testgles2 "test/testgles2.c")
-  test_program(testgles2_sdf "test/testgles2_sdf.c;test/testutils.c")
+  if(HAVE_OPENGLES)
```
Comment (sezero) [inline CMakeLists.txt:374]: This should be `HAVE_OPENGLES2`. `HAVE_OPENGLES` check itself should be added to `testgles`.

File: CMakeLists.txt [hunk 4]
```diff
@@ -22,7 +22,20 @@ include(CheckCSourceCompiles)
 include(CheckIncludeFile)
 include(CheckCCompilerFlag)
 include(CMakePackageConfigHelpers)
+include(CMakePushCheckState)
 include(GNUInstallDirs)
+include("${CMAKE_CURRENT_LIST_DIR}/cmake/sdlchecks.cmake")
```
Comment (sezero) [inline CMakeLists.txt:27]: This isn't anywhere. (forgot to `git add` ?)

File: CMakeLists.txt (hunk 2 above)
Comment (madebr) [reply CMakeLists.txt:32]: SDL does only add `-Wl,--no-undefined` for non-Apple non-OpenBSD platforms.
Shouldn't apple also get it?

Comment (sezero) [reply CMakeLists.txt:32]: For Apple's ld, we use `-Wl,-undefined,error` (someone recently removed it for reasons I don't fully understand.)

Comment (madebr) [reply CMakeLists.txt:32]: https://github.com/libsdl-org/SDL/pull/6365
It's redundant, he says.

File: CMakeLists.txt (hunk 4 above)
Comment (madebr) [reply CMakeLists.txt:27]: Indeed, git equivalent of forgot to add the email attachment :)

File: CMakeLists.txt (hunk 2 above)
Comment (sezero) [reply CMakeLists.txt]: Hm. Ookay. (I still have it in my standalone Makefile.darwin)

File: CMakeLists.txt (hunk 3 above)
Comment (madebr) [reply CMakeLists.txt:378]: Indeed, fix this.

File: CMakeLists.txt (hunk 1 above)
Comment (madebr) [reply CMakeLists.txt]: Will do.

File: CMakeLists.txt [hunk 5]
```diff
@@ -209,11 +226,18 @@ endif()
 #else()
     add_library(SDL2main STATIC src/SDLmain/dummy/SDL_dummy_main.c)
 #endif()
-#add_library(SDL2::SDL2ain ALIAS SDL2main)
+add_library(SDL2::SDL2main ALIAS SDL2main)
 target_include_directories(SDL2main PRIVATE "include/SDL2")
+if(MINGW OR CYGWIN)
+  if(CMAKE_SIZEOF_VOID_P EQUAL 4)
```
Comment (sezero) [inline CMakeLists.txt:232]: I wonder whether the `_` prefix and `@16` stdcall suffix applies to 32 bit targets other than x86, e.g. 32 bit arm...

**Comment by sezero** [comment]
Merging this as is now.

**Pull request merged**